staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -697,33 +696,25 @@ func (p *processorListener) pop() {
 
 func (p *processorListener) run() {
 	// this call blocks until the channel is closed.  When a panic happens during the notification
-	// we will catch it, **the offending item will be skipped!**, and after a short delay (one minute)
+	// we will catch it, **the offending item will be skipped!**, and after a delay (one minute)
 	// the next notification will be attempted.  This is usually better than the alternative of never
 	// delivering again.
 	stopCh := make(chan struct{})
 	wait.Until(func() {
-		// this gives us a few quick retries before a long pause and then a few more quick retries
-		err := wait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
-			for next := range p.nextCh {
-				switch notification := next.(type) {
-				case updateNotification:
-					p.handler.OnUpdate(notification.oldObj, notification.newObj)
-				case addNotification:
-					p.handler.OnAdd(notification.newObj)
-				case deleteNotification:
-					p.handler.OnDelete(notification.oldObj)
-				default:
-					utilruntime.HandleError(fmt.Errorf("unrecognized notification: %T", next))
-				}
+		for next := range p.nextCh {
+			switch notification := next.(type) {
+			case updateNotification:
+				p.handler.OnUpdate(notification.oldObj, notification.newObj)
+			case addNotification:
+				p.handler.OnAdd(notification.newObj)
+			case deleteNotification:
+				p.handler.OnDelete(notification.oldObj)
+			default:
+				utilruntime.HandleError(fmt.Errorf("unrecognized notification: %T", next))
```

Comment:
I don't see how this actually causes a delay as claimed by the comment? Maybe the original author intended to have a `return`?

Also, doesn't a minute seem overly punitive? It's already skipping the item, why also delay?

I think the minute on the outer Wait was there to make sure it would lose the race with the `close(stopCh)` call, and the inner wait was intended to do exponential backoffs. But the original author missed a `return` or thought that HandleError panic'd, and as a result there actually was no delay.

Reply:
`wait.Until` does the catch and delay, see the amount of delay is the penultimate parameter.

So should the `wait.Until` be replaced with an exponential backoff?  If so, when would it reset?

Reply:
`wait.ExponentialBackoff` does not catch and handle panics, so this PR is making no change in behavior.  But we are getting deep into questions of original intent and (more importantly, since the original code looks confused) what we think _should_ be done.  So let's do this.  We are not talking about retrying a notification, we are talking about proceeding from one notification to the next when processing the first notification panics.  What pattern of delays do we want?  Do we have established principles for answering this question?  A panic indicates a bug (perhaps one of several).  But unless we add significant smarts, this code will not know how often these bugs cause a panic.  I am inclined to suggest a constant short delay, perhaps one second.  Short should be OK because each attempt makes progress (consumes at least one item from the queue).  Longer could be bad because it introduces unhelpful delay when a bug bites.  But the delay should be long enough that something in the context can plausibly have changed, because the bugs may depend on something that changes with time.

Reply:
Ah, didn't realize ExponentialBackoff doesn't catch panics. Yeah, this inner func can never return an error, so the inner backoff really was useless.

I agree that a minute seems uselessly long here. ISTM there are two reasonably likely cases:
1. a single notification tickles some bug in one-off fashion
1. some incompatibility means *every* notification will cause a panic, leading to a flood of panic-recover cycles

I think you're right that once a second strikes a reasonable balance between not suppressing problems that need to be addressed, and not overwhelming the client program with a potentially-infinite flood of error logging.

Reply:
(BTW, this is likely moot since `runtime.ReallyCrash` defaults to true these days, so it looks like the first panic will just kill the program most of the time.)

Reply:
HandleError doesn't panic. And it already does backoff: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L88

Reply:
I may have been confused, and/or the original comment wasn't clear.

* This error delivery line will skip an item. But no panic is involved.
* A panic on an item will not cause this line to be executed. It will also not cause the item to be "skipped", since the item was in the middle of processing in order to cause the panic. It won't be re-delivered/retried, but that's not the same as skipping it. It might have gotten 99.9% done and made irreversible state changes for all we know.

Reply:
I think a fixed delay is fine, since we are not retrying the same item (nor do we want to, since it caused a panic!). I think 60 seconds is too long and is likely to cause the channel to fill up or get too big. I think 1s is enough to avoid a hot loop.

I also think ReallyCrash should be on for most callers, so this doesn't matter too much.

Who knows what state the recipient might be in after panicing partway through processing some item!

Reply:
While this discussion is attached to the default case of the switch, the subject matter moved to focus on the comment at the start of the function body.
I think we have agreed what to do, and I have pushed that revision.